src/configuration.c

```diff
@@ -0,0 +1,202 @@
+	action_t action = action_by_name(value);
+
+	if (strlen(key) == 3 && (key[0] == 'c' || key[0] == 'C') && key[1] == '-') {
+		const char letter = toupper(key[2]);
+		if (letter < 'A' || letter > 'Z') {
+			fprintf(stderr, "%s control code invalid, must be a letter from a to z\n", key);
+			exit(1);
+		}
+
+		char ctrl_code[2];
+		ctrl_code[0] = letter - '@';
+		ctrl_code[1] = '\0';
+		if (insert_keybinding(keybindings, keybinding_count, ctrl_code, action)) {
+			++keybinding_count;
+		}
+	}
+	else if (! strcmp(key, "up")) {
+		if (insert_keybinding(keybindings, keybinding_count, "\x1b[A", action)) {
+			++keybinding_count;
+		}
+		if (insert_keybinding(keybindings, keybinding_count, "\x1bOA", action)) {
+			++keybinding_count;
+		}
+	}
+	else if (! strcmp(key, "down")) {
+		if (insert_keybinding(keybindings, keybinding_count, "\x1b[B", action)) {
+			++keybinding_count;
+		}
+		if (insert_keybinding(keybindings, keybinding_count, "\x1bOB", action)) {
+			++keybinding_count;
+		}
+	}
+	else if (! strcmp(key, "page up")) {
+		if (insert_keybinding(keybindings, keybinding_count, "\x1b[5~", action)) {
+			++keybinding_count;
+		}
+	}
+	else if (! strcmp(key, "page down")) {
+		if (insert_keybinding(keybindings, keybinding_count, "\x1b[6~", action)) {
+			++keybinding_count;
+		}
+	}
+	else if (! strcmp(key, "delete")) {
+		if (insert_keybinding(keybindings, keybinding_count, "\x7f", action)) {
+			++keybinding_count;
+		}
+	}
+	else if (! strcmp(key, "tab")) {
+		if (insert_keybinding(keybindings, keybinding_count, KEY_CTRL('I'), action)) {
+			++keybinding_count;
+		}
+	}
+	else {
+			fprintf(stderr, "%s it not a valid keybinding\n", key);
+			exit(1);
+	}
+
+	return keybinding_count;
+}
+#endif
+
+configuration_t *configuration_init() {
+	configuration_t *configuration = safe_realloc(NULL, sizeof(configuration_t));
+	size_t keybindings_count = sizeof(default_keybindings) / sizeof(indexed_action_t);
+
+	configuration->keybindings = (keybinding_t *) safe_realloc(NULL, (keybindings_count + 1) * sizeof(keybinding_t));
+	for (size_t i = 0; i < keybindings_count; i++) {
+		keybinding_t *binding = configuration->keybindings + i;
+		binding->key = safe_strdup(default_keybindings[i].key);
+		binding->action = default_keybindings[i].action;
+	}
+
+#ifdef WITH_LIBCONFIG
+	const char *fzy_cfg_path = xdgConfigFind("fzy", NULL);
+	if (strlen(fzy_cfg_path) > 0) {
+		config_t cfg;
+		config_init(&cfg);
+		if (!config_read_file(&cfg, fzy_cfg_path)) {
+			fprintf(stderr, "could not read config %s:%d - %s\n", config_error_file(&cfg), config_error_line(&cfg), config_error_text(&cfg));
+			exit(1);
+		}
+
+		config_setting_t *setting = config_lookup(&cfg, "key-bindings");
+		if (setting) {
+			int setting_count = config_setting_length(setting);
+			configuration->keybindings = (keybinding_t *) safe_realloc(
+				configuration->keybindings,
+				(keybindings_count + setting_count + 1) * sizeof(keybinding_t)
```

Comment:
Allocate enough space to cover the user adding only new keybindings. Since "up" and "down" are in the default set they won't end up allocating two extra spaces, otherwise we'd need to `+ 3` here.